java/foreign-bridge/src/main/java/sleeper/foreign/bridge/FFIContext.java

```diff
@@ -48,22 +53,46 @@ public class FFIContext<T extends ForeignFunctions> implements AutoCloseable {
      * Pointer to the Rust side of the FFI layer. If this is null, it means the
      * context has been closed.
      */
-    private Pointer context;
+    private volatile Pointer context;
 
     /**
      * Initialises the FFI library and context for calling functions.
      *
      * This will attempt to extract the native library from the JAR file and
      * load it into the JVM. It will then establish the Rust side of the context
-     * to enable queries to be executed.
+     * to enable FFI calls to be executed.
+     *
+     * @param  functionClass        the native function interface type
+     * @throws NullPointerException if any parameter is null
+     * @throws IOException          if the native library couldn't be loaded
+     */
+    public FFIContext(Class<T> functionClass) throws IOException {
+        this.ffiClass = Objects.requireNonNull(functionClass, "functionClass must not be null");
+        this.functions = FFIBridge.createForeignInterface(functionClass);
+        this.context = Objects.requireNonNull(functions.create_context(), "FFI create_context returned null");
+    }
+
+    /**
+     * Clones this context from an existing one.
+     *
+     * This is useful for creating a new context instance from an existing one to use on a new thread.
      *
-     * @param functions the native function interface
+     * @param  original              the context to clone
+     * @throws IOException           if the native library couldn't be loaded
+     * @throws IllegalStateException if the original context has already been closed
      */
-    public FFIContext(T functions) {
-        this.functions = Objects.requireNonNull(functions, "functions must not be null");
-        // Create Java interface to FFI lib
-        // Make FFI call to establish foreign context
-        this.context = Objects.requireNonNull(functions.create_context(), "context must not be null");
+    public FFIContext(FFIContext<T> original) throws IOException {
+        original.checkClosed();
```

Comment:
This reads as though we're checking that the original context is closed, when we're actually checking it's still open. Please can we rename this method?